### PR TITLE
Add softness setting

### PR DESCRIPTION
**Pull request by CandyFace**
The new setting allows for softer looking brushes, by adjusting the opacity and the factor, whereas the offset was previously constant below a certain level.

Depending on how you combine softness and hardness, you can get some pretty interesting results. 
My motivation here was to be able to produce softer looking brushes, with more ease. 
Screenshot 2020-01-11 at 18 16 18
Screenshot 2020-01-11 at 18 04 55
Screenshot 2020-01-11 at 18 15 01

hardness and softness low, same values at various levels
Screenshot 2020-01-11 at 18 20 21

Hardness at 0.5 and softness increasing from 0 to 0.99.
image

### CODE REVIEW

File: brushsettings.json [hunk 1]
```diff
@@ -232,6 +232,15 @@
             "minimum": 0.0,
             "tooltip": "Hard brush-circle borders (setting to zero will draw nothing). To reach the maximum hardness, you need to disable Pixel feather."
         },
+        {
+            "constant": false,
+            "default": 0.5,
```
Comment (jplloyd) [inline brushsettings.json]: This should be set to 0.0 to not break existing brushes.

**Comment by jplloyd** [comment]
Alright, so this is basically a linear scaling for the slope and offset values calculated from the hardness parameter.

There are a couple of questions we need to answer here:

1. Does this do anything that cannot be done using only existing settings? This is not a rhetorical question. I don't know enough about the interaction of all parameters to say for sure. Even if it is not the case, it would still be worth including if it makes brush creation easier/more intuitive.

2. Does this break any assumptions about the way the hardness value works? Again, this is not an area where I know the code and equations in depth. I want to be sure about this before merging.

----

Practicalities.

Since this won't be merged into the 1.5 release (string freeze and all that), it's not that important, but it should be noted that these changes are API- and ABI-breaking, due to the changes to public function signatures and the insertion of the new setting in the middle of the existing settings (the order in the json file determines the order in the enums). Again, since the API/ABI for 2.0 is not locked down (and won't be for a while, due to the 1.5 backporting effort), it doesn't really matter, but is something to think about for the future.

Additionally, the default value for a new setting should be set so that it does not affect existing brushes  - in this case the default value should be 0 and not 0.5.

---

I don't have time to test this, or verify that this change doesn't break something, until the MyPaint 2.0 release, but if there are no issues it'll definitely be merged for libmypaint 2.0.

File: brushsettings.json (hunk 1 above)
Comment (CandyFace) [reply brushsettings.json]: Ah.. yes, fixed.

**Comment by CandyFace** [comment]
> Does this do anything that cannot be done using only existing settings? This is not a rhetorical question. I don't know enough about the interaction of all parameters to say for sure. Even if it is not the case, it would still be worth including if it makes brush creation easier/more intuitive.

A combination of tweaking hardness and opaque and possibly other settings would be able to achieve some of the above, but the last image, which produces a variety of smooth strokes where opacity blends more linearly from center to edge is something I haven't been able to reproduce without softness setting.

> Does this break any assumptions about the way the hardness value works? Again, this is not an area where I know the code and equations in depth. I want to be sure about this before merging.

I've tested it a fair bit and have yet to crash my application, precautions have been made to avoid softness becoming 1, so I'd like to say it doesn't but I can't be 100% sure.

**Comment by jplloyd** [comment]
Rebased, resolved, and squashed here: 49357432eafced4b0dc798f0d80ad99b93b4ad99

**Pull request closed without merging**